### ISSUE TITLE
Nvwave, Outlook/UIA, sometimes Chromium Edge, NVDA freezes and needs to forcefully be restarted

### ISSUE THREAD

**Issue by MarcoZehe** [nvaccess/nvda#11576]
### Steps to reproduce:

This was a freeze I encountered after closing a message in Outlook 365 and returning to the messages list. NVDA fell silent, braille display froze, but error sounds were heard sporadically. I had to forcefully restart NVDA via its hot key to get speech and braille back. It did manage to save the configuration, though.

The frozen periods resulted in the following log entries:

```
ERROR - watchdog._watcher (17:42:10.082) - watchdog (22120):
Core frozen in stack!
INFO - watchdog._watcher (17:42:10.389) - watchdog (22120):
Listing stacks for Python threads:
Python stack for thread 26508 (nvwave.playWaveFile(error.wav)):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "nvwave.pyc", line 417, in idle
  File "nvwave.pyc", line 427, in _idleUnbuffered
  File "nvwave.pyc", line 377, in sync
  File "winKernel.pyc", line 225, in waitForSingleObject

Python stack for thread 25672 (watchdog.CancellableCallThread.execute(<_FuncPtr object at 0x00BD7BE8>)):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "watchdog.pyc", line 317, in run
  File "threading.pyc", line 552, in wait
  File "threading.pyc", line 296, in wait

Python stack for thread 22120 (watchdog):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "watchdog.pyc", line 127, in _watcher
  File "watchdog.pyc", line 62, in getFormattedStacksForAllThreads

Python stack for thread 24116 (touchHandler.TouchHandler):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "touchHandler.pyc", line 241, in run

Python stack for thread 31100 (winInputHook):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "winInputHook.pyc", line 79, in hookThreadFunc

Python stack for thread 22516 (_UIAHandler.UIAHandler.MTAThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "_UIAHandler.pyc", line 310, in MTAThreadFunc
  File "queue.pyc", line 170, in get
  File "threading.pyc", line 296, in wait

Python stack for thread 30740 (braille._BgThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "braille.pyc", line 2187, in func

Python stack for thread 12740 (Thread-1):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "C:\Users\marco\AppData\Roaming\nvda\addons\vocalizer-driver\synthDrivers\vocalizer\_vocalizer.py", line 36, in run
    func(*args, **kwargs)
  File "C:\Users\marco\AppData\Roaming\nvda\addons\vocalizer-driver\synthDrivers\vocalizer\_vocalizer.py", line 279, in _processText2Speech
    vautoDll.vauto_ttsProcessText2Speech(instance, byref(inText))
  File "C:\Users\marco\AppData\Roaming\nvda\addons\vocalizer-driver\synthDrivers\vocalizer\_vocalizer.py", line 96, in callback
    player.feed(data[prevByte:])
  File "nvwave.pyc", line 317, in feed
  File "nvwave.pyc", line 332, in _feedUnbuffered_handleErrors
  File "nvwave.pyc", line 360, in _feedUnbuffered
  File "nvwave.pyc", line 377, in sync
  File "winKernel.pyc", line 225, in waitForSingleObject

Python stack for thread 25908 (MainThread):
  File "nvda.pyw", line 215, in <module>
  File "core.pyc", line 550, in main
  File "wx\core.pyc", line 2134, in MainLoop
  File "gui\__init__.pyc", line 1050, in Notify
  File "core.pyc", line 519, in run
  File "IAccessibleHandler\__init__.pyc", line 1085, in pumpAll
  File "IAccessibleHandler\__init__.pyc", line 759, in processFocusWinEvent
  File "IAccessibleHandler\__init__.pyc", line 636, in winEventToNVDAEvent
  File "NVDAObjects\IAccessible\__init__.pyc", line 42, in getNVDAObjectFromEvent
  File "IAccessibleHandler\__init__.pyc", line 412, in accessibleObjectFromEvent
  File "oleacc.pyc", line 265, in AccessibleObjectFromEvent

ERROR - watchdog._watcher (17:42:25.130) - watchdog (22120):
Core frozen in stack!
INFO - watchdog._watcher (17:42:25.428) - watchdog (22120):
Listing stacks for Python threads:
Python stack for thread 30344 (nvwave.playWaveFile(error.wav)):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "nvwave.pyc", line 417, in idle
  File "nvwave.pyc", line 427, in _idleUnbuffered
  File "nvwave.pyc", line 377, in sync
  File "winKernel.pyc", line 225, in waitForSingleObject

Python stack for thread 25672 (watchdog.CancellableCallThread.execute(<_FuncPtr object at 0x00BD7BE8>)):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "watchdog.pyc", line 317, in run
  File "threading.pyc", line 552, in wait
  File "threading.pyc", line 296, in wait

Python stack for thread 22120 (watchdog):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "watchdog.pyc", line 127, in _watcher
  File "watchdog.pyc", line 62, in getFormattedStacksForAllThreads

Python stack for thread 24116 (touchHandler.TouchHandler):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "touchHandler.pyc", line 241, in run

Python stack for thread 31100 (winInputHook):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "winInputHook.pyc", line 79, in hookThreadFunc

Python stack for thread 22516 (_UIAHandler.UIAHandler.MTAThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "_UIAHandler.pyc", line 310, in MTAThreadFunc
  File "queue.pyc", line 170, in get
  File "threading.pyc", line 296, in wait

Python stack for thread 30740 (braille._BgThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "braille.pyc", line 2187, in func

Python stack for thread 12740 (Thread-1):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "C:\Users\marco\AppData\Roaming\nvda\addons\vocalizer-driver\synthDrivers\vocalizer\_vocalizer.py", line 36, in run
    func(*args, **kwargs)
  File "C:\Users\marco\AppData\Roaming\nvda\addons\vocalizer-driver\synthDrivers\vocalizer\_vocalizer.py", line 279, in _processText2Speech
    vautoDll.vauto_ttsProcessText2Speech(instance, byref(inText))
  File "C:\Users\marco\AppData\Roaming\nvda\addons\vocalizer-driver\synthDrivers\vocalizer\_vocalizer.py", line 96, in callback
    player.feed(data[prevByte:])
  File "nvwave.pyc", line 317, in feed
  File "nvwave.pyc", line 332, in _feedUnbuffered_handleErrors
  File "nvwave.pyc", line 360, in _feedUnbuffered
  File "nvwave.pyc", line 377, in sync
  File "winKernel.pyc", line 225, in waitForSingleObject

Python stack for thread 25908 (MainThread):
  File "nvda.pyw", line 215, in <module>
  File "core.pyc", line 550, in main
  File "wx\core.pyc", line 2134, in MainLoop
  File "gui\__init__.pyc", line 1050, in Notify
  File "core.pyc", line 519, in run
  File "IAccessibleHandler\__init__.pyc", line 1085, in pumpAll
  File "IAccessibleHandler\__init__.pyc", line 759, in processFocusWinEvent
  File "IAccessibleHandler\__init__.pyc", line 636, in winEventToNVDAEvent
  File "NVDAObjects\IAccessible\__init__.pyc", line 42, in getNVDAObjectFromEvent
  File "IAccessibleHandler\__init__.pyc", line 412, in accessibleObjectFromEvent
  File "oleacc.pyc", line 265, in AccessibleObjectFromEvent

ERROR - watchdog._watcher (17:42:40.186) - watchdog (22120):
Core frozen in stack!
INFO - watchdog._watcher (17:43:26.887) - watchdog (22120):
Listing stacks for Python threads:
Python stack for thread 30384 (nvwave.playWaveFile(error.wav)):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "nvwave.pyc", line 417, in idle
  File "nvwave.pyc", line 427, in _idleUnbuffered
  File "nvwave.pyc", line 377, in sync
  File "winKernel.pyc", line 225, in waitForSingleObject

Python stack for thread 25672 (watchdog.CancellableCallThread.execute(<_FuncPtr object at 0x00BD7BE8>)):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "watchdog.pyc", line 317, in run
  File "threading.pyc", line 552, in wait
  File "threading.pyc", line 296, in wait

Python stack for thread 22120 (watchdog):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "watchdog.pyc", line 127, in _watcher
  File "watchdog.pyc", line 62, in getFormattedStacksForAllThreads

Python stack for thread 24116 (touchHandler.TouchHandler):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "touchHandler.pyc", line 241, in run

Python stack for thread 31100 (winInputHook):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "winInputHook.pyc", line 79, in hookThreadFunc

Python stack for thread 22516 (_UIAHandler.UIAHandler.MTAThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "_UIAHandler.pyc", line 310, in MTAThreadFunc
  File "queue.pyc", line 170, in get
  File "threading.pyc", line 296, in wait

Python stack for thread 30740 (braille._BgThread):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "threading.pyc", line 870, in run
  File "braille.pyc", line 2187, in func

Python stack for thread 12740 (Thread-1):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "C:\Users\marco\AppData\Roaming\nvda\addons\vocalizer-driver\synthDrivers\vocalizer\_vocalizer.py", line 32, in run
    func, args, kwargs = bgQueue.get()
  File "C:\Users\marco\AppData\Roaming\nvda\addons\vocalizer-driver\synthDrivers\vocalizer\_vocalizer.py", line 282, in _processText2Speech
    speakingInstance = None

Python stack for thread 25908 (MainThread):
  File "nvda.pyw", line 215, in <module>
  File "core.pyc", line 550, in main
  File "wx\core.pyc", line 2134, in MainLoop
  File "gui\__init__.pyc", line 1052, in Notify
  File "core.pyc", line 531, in run
  File "queueHandler.pyc", line 88, in pumpAll
  File "queueHandler.pyc", line 57, in flushQueue
  File "watchdog.pyc", line 77, in alive
```

### Actual behavior:

Freeze, see above.

### Expected behavior:

No freeze.

#### NVDA installed/portable/running from source:

Installed.

#### NVDA version:

NVDA version alpha-20851,f33664ec

#### Windows version:

20H2.

#### Name and version of other software in use when reproducing the issue:

Outlook 365.

#### Other information about your system:

Focus 40 Blue 5th Generation Braille display used.

#### Does the issue still occur after restarting your computer?

Unknown, not reproducible.

#### Have you tried any other versions of NVDA? If so, please report their behaviors.

No.

#### If addons are disabled, is your problem still occuring?

Probably, if it happens.

#### Did you try to run the COM registry fixing tool in NVDA menu / tools?

No.

**Comment by akash07k** [comment]
Ya, it is happening in alpha releases with me too. sometimes
encountered it on firefox too

**Comment by michaelDCurran** [comment]
It looks like that NVDA's main thread is freezing in a call to AccessibleObjectFromEvent when trying to handle a MSAA focus event. 
nvWave is mentioned in these stacks, but I think that is only because the error sound has just been played due to the logging of the core being frozen. I don't think that is part of the freeze necessarily. All the other threads just seem to be sitting happily in their weight states for something to do.

The only questionable one is the vocalizer thread. The first time, it looks like it is waiting on nvWave for some audio to be finished playing. I'm not quite sure what this could be as I would have thought that any audio would havefinished a long time before the core froze... unless it really was speaking for a good 15 seconds?
Later on there is another call stack for a Vocalizer thread, which I'm rather confused about:
```
Python stack for thread 12740 (Thread-1):
  File "threading.pyc", line 890, in _bootstrap
  File "threading.pyc", line 926, in _bootstrap_inner
  File "C:\Users\marco\AppData\Roaming\nvda\addons\vocalizer-driver\synthDrivers\vocalizer\_vocalizer.py", line 32, in run
    func, args, kwargs = bgQueue.get()
  File "C:\Users\marco\AppData\Roaming\nvda\addons\vocalizer-driver\synthDrivers\vocalizer\_vocalizer.py", line 282, in _processText2Speech
    speakingInstance = None
```
This looks like some of is code is reentered inside fetching from its queue. I've never quite seen that before. Assuming they aren't using any task feachers of queue.queue, they may consider switching to queue.SimpleQueue, to remove any possibility of code being executed within the fetch... but I don't think this is again related to the freeze.
Interestingly, the very last list of call stacks shows that the core is starting to become alive again. I'm not sure how it broke out of the freeze... just luck perhaps.

In short though, I think that Outlook, or what ever the focus was fore, simply froze itself. It would be nice of AccessibleObjectFromEvent would not freeze... and in deed it really shouldn't -- we have multiple mechanisms to try and cancel this. But clearly none of them are enough.
I don't think this is new for 2020.3. But if you can find a way of reproducing this, please provide another log, but also with the MSAA debug logging enabled in NVDA's Advanced settings.

@akash07k  If you also can provide a log showing the Firefox one that would be very useful.

**Comment by MarcoZehe** [comment]
I have thus far intermittently seen this in more apps, such as Firefox, Chromium Edge, and even Reaper once. So this doesn't seem UIA and Outlook specific after all, it was just the first time I saw it, that it was in Outlook. I have also seen this with alpha versions on two machines now. But again, no exact steps to reproduce, it happens intermittently.

**Comment by MarcoZehe** [comment]
I am still seeing this in latest alpha versions. I also see it sometimes now when waking the machine and unlocking, and the last window is a browser window, e.g . Chromium Edge, either as the browser or when running a PWA. The symptoms are the same: NVDA loses speech, Braille freezes, and every 20 seconds or so, an error sound is heard. I need to forcefully restart via the Desktop shortcut to get it back. I don't hear the shutdown sound, just the startup sound from the new instance.

Happens on two very different machines (hardware-wise), with two brands of Braille display connected, which both still work when unlocking the machines.

Maybe related: #11723

**Comment by michaelDCurran** [comment]
As annoying as this is, each time that happens, can you please collect the log and provide it here?  It would be good to know if the call stacks always look the same. Every time you here the error sound every 15 - 20 seconds, NVDA has detected the core is frozen and is printing the stacks to the log.

**Comment by michaelDCurran** [comment]
Could you also consider setting your log level to debug, and also enabling the MSAA checkbox in the Debug log section of NVDA's advanced settings panel. There may be a slight performance decrease when doing this, but this is the best way we can get further information. I'd really like to know what winEvent AccessibleObjectFromEvent is freezing on.

**Comment by MarcoZehe** [comment]
Here we go! A full log from the moment I woke the machine until I pressed CTRL+Alt+N after hearing the first error sound from WatchDog. Unzipped, this is 400 KB worth of logs.
[NVDA ran into freeze after waking machine.zip](https://github.com/nvaccess/nvda/files/5338321/NVDA.ran.into.freeze.after.waking.machine.zip)

**Comment by MarcoZehe** [comment]
Here's another one, this one happened right after Alpha installed the October 6 update. Unfortunately, it didn't save the logging changes from the last session, and I forgot to save it manually yesterday. So this one doesn't contain MSAA logging. But it should still contain the stack traces.
[nvda ran into freeze after update.zip](https://github.com/nvaccess/nvda/files/5338330/nvda.ran.into.freeze.after.update.zip)

**Comment by michaelDCurran** [comment]
NvDA ran into freeze after waking:

* I can see that you were running MS Edge.
* NVDA got a focus MSAA event for a Twitter document.
* But after that, NVDA seems to be processing a UI Automation focus event for some kind of edit field or editable document (might be the Twitter document)

NvDA's main thread got into some kind of endless recursion trying to gather text via UIAutomation:
```
Python stack for thread 604 (MainThread):
  File "nvda.pyw", line 236, in <module>
  File "core.pyc", line 553, in main
  File "wx\core.pyc", line 2134, in MainLoop
  File "gui\__init__.pyc", line 1062, in Notify
  File "core.pyc", line 523, in run
  File "queueHandler.pyc", line 88, in pumpAll
  File "queueHandler.pyc", line 55, in flushQueue
  File "eventHandler.pyc", line 65, in _queueEventCallback
  File "eventHandler.pyc", line 220, in executeEvent
  File "eventHandler.pyc", line 96, in __init__
  File "eventHandler.pyc", line 105, in next
  File "NVDAObjects\behaviors.pyc", line 187, in event_gainFocus
  File "NVDAObjects\UIA\__init__.pyc", line 1067, in event_gainFocus
  File "NVDAObjects\__init__.pyc", line 1108, in event_gainFocus
  File "NVDAObjects\__init__.pyc", line 994, in reportFocus
  File "speech\__init__.pyc", line 478, in speakObject
  File "speech\__init__.pyc", line 548, in getObjectSpeech
  File "speech\types.pyc", line 50, in _flattenNestedSequences
  File "speech\types.pyc", line 50, in <genexpr>
  File "speech\__init__.pyc", line 1127, in getTextInfoSpeech
  File "NVDAObjects\UIA\__init__.pyc", line 716, in getTextWithFields
  File "NVDAObjects\UIA\__init__.pyc", line 686, in _getTextWithFieldsForUIARange
  File "NVDAObjects\UIA\__init__.pyc", line 686, in _getTextWithFieldsForUIARange
  File "NVDAObjects\UIA\__init__.pyc", line 686, in _getTextWithFieldsForUIARange
  [Previous line repeated 39 more times]
  File "comtypesMonkeyPatches.pyc", line 26, in __call__
```
Or perhaps that final uIA call is actually freezing -- hard to tell.
The last time I tested Chromeium's UIA implemenation (in 2019) there were definitely freezes and or strange textRanges that caused NvDA to infinitly recurse.

The big question here however is why NVDA processed both a MSAA focus event and a UI  Automation focus event.
We mark the Chrome_RenderWidgetHostHWND window class as being a bad UIA window (meaning that only MSAA should be used). 
In this case, NvDA got it right and allowed the MSAA event through, but got it wrong and did not deny the UIA event.
Some how UIAHandler.handler.getNearestWindowHandle located the wrong window handle and just happened to find one that definitely had a native UIA implementation, which we did not mark as bad...
Perhaps this only happens when resuming from sleep... I shall investigate further.

**Comment by michaelDCurran** [comment]
NvDA froze after update:
Not enough to go on sadly.
NvDA's main thread froze when trying to fetch states for a UIA NVDAObject.
Were you by chance also running Edge here?

**Comment by MarcoZehe** [comment]
Yes, Edge was the foreground window in this case as well.

**Comment by michaelDCurran** [comment]
Might you be able to provide the exact version number for your copy of 
Edge please?

**Comment by MarcoZehe** [comment]
Version 85.0.564.68 (Official Build) (64-Bit). There are no configuration changes via Experiments or such to force UIA or anything.

This all started happening around the build I initially filed this issue with. I had not seen it before that or the snapshot build immediately prior to that. Perhaps that gives a hint as to what could have introduced this.

**Comment by michaelDCurran** [comment]
I was able to reproduce the situation where resuming / unlocking, focus goes to document content in Edge, but uses UIA instead of MSAA.
This can also be reproduced by focusing some document content, then  opening the context menu (with the applications key) and then pressing escape to dismiss it.
Related issues: #11202 and #11536 

The bug is as follows:
* The object (or Window) that had focus is destryed, but NVDA still considers it the focused object as there has not been a new focus event yet.
* NVDA tries to locate the actual focus by asking the OS for the window with focus.
* This window apparently natively supports UI Automation. In this case it is a "Chrome_WidgetWin_1" window.
* In NVDAObjects.UIA.UIA.kwargsFromSuper: NVDA asks the OS for the UI Automation element with focus. In this case it gets back a link or other document content from Edge.
* NVDA however does not bother checking if this UIA element is a usable UIA element (I.e. it is not proxied, and we have not marked its window as being bad etc).

Like with similar code that asks the OS for an object at particular hittesting coordinates, it should ensure the UIA element it gets back is usable (UIAhandler.handler.isNativeUIAElement) and if it is not, it should return False, and thus Window NVDAObject will go on and try the focus according to MSAA.

I shall file a PR for this.

Note however, that this only deals with the problems in this issue related to resuming / unlocking in MS Edge. The freezes with AccessibleObjectFromEvent are unrelated.

**Comment by michaelDCurran** [comment]
@MarcoZehe  I believe I addresed the freeze when focus lands in MS Edge after resuming / unlocking the machine. This pr has been in master for a few days. Can you confirm that the problem has been solved for you? 
The first freeze that this issue started with in AccessibleObjectFromEvent is different, and I would probably suggest that if you see that one again, that you file a separate issue, and we treat this one as just dealing with the MS Edge freeze, which hopefully is fixed.

**Comment by MarcoZehe** [comment]
I can confirm that that freeze is fixed for me, too. Moreover, I have only seen that freeze in Outlook once, while the MS Edge freezes happened regularly. I will close this as fixed, and if I ever see that freeze in Outlook again, I'll open a new issue.
